crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java

```diff
@@ -0,0 +1,77 @@
+package edu.uci.ics.crawler4j.crawler.dynamic;
+
+import edu.uci.ics.crawler4j.crawler.DynamicCrawlConfig;
+import edu.uci.ics.crawler4j.crawler.WebCrawler;
+import edu.uci.ics.crawler4j.parser.dynamic.DynamicParser;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.chrome.ChromeDriver;
+import org.openqa.selenium.chrome.ChromeOptions;
+import org.openqa.selenium.firefox.FirefoxDriver;
+import org.openqa.selenium.firefox.FirefoxOptions;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Extends {@link WebCrawler} for using a Selenium {@link WebDriver} to fetch dynamically generated HTML pages.
+ */
+public class DynamicWebCrawler extends WebCrawler {
+
+    private WebDriver webDriver;
+
+    @Override
+    public void onStart() {
+        // Instantiate and register the WebDriver instance for this crawler thread to the parser.
+        // This is due to the fact Selenium's WebDriver is not thread-safe.
+        webDriver = newWebDriverInstance((DynamicCrawlConfig) getMyController().getConfig());
+        ((DynamicParser) getMyController().getParser()).setWebDriver(webDriver);
+    }
+
+    @Override
+    public void onBeforeExit() {
+        this.webDriver.close();
+    }
+
+    /**
+     * Default options for the headless browser
+     * @param config the {@link DynamicCrawlConfig} where the user agent is defined.
+     * @return the default options for the headless browser
+     */
+    public List<String> defaultHeadlessBrowserOptions(DynamicCrawlConfig config) {
+        return Arrays.asList(
```

Comment:
What happens if an options is specified multiple times, i.e. a user puts in `--window-size=800,600`. Which option has preference? Should we check for duplicates and replace?

Reply:
That's not great indeed. I've replaced the `List` with a `Map<String, Optional<String>> ` and now options defined in the crawl config will override default ones.